ruby/ql/lib/upgrades/9fdd1d40fd3c3f8f9db8fabf5a353580d14c663a/ruby_ast_node_info.ql

```diff
@@ -0,0 +1,45 @@
+class ScopeResolutionMethodCall extends @ruby_call {
+  private @ruby_scope_resolution scope;
+
+  ScopeResolutionMethodCall() { ruby_call_def(this, scope) }
+
+  @ruby_scope_resolution getScopeResolution() { result = scope }
+
+  string toString() { none() }
+}
+
+class Location extends @location {
+  string toString() { none() }
+}
+
+class RubyAstNode extends @ruby_ast_node {
+  string toString() { none() }
+}
+
+class RubyAstNodeParent extends @ruby_ast_node_parent {
+  string toString() { none() }
+}
+
+from RubyAstNode node, RubyAstNodeParent parent, int index, Location loc
+where
+  ruby_ast_node_info(node, parent, index, loc) and
+  not node = any(ScopeResolutionMethodCall c).getScopeResolution() and
+  not parent = any(ScopeResolutionMethodCall c) and
+  not parent = any(ScopeResolutionMethodCall c).getScopeResolution()
+  or
+  ruby_ast_node_info(node, _, _, loc) and
+  parent instanceof ScopeResolutionMethodCall and
+  node =
+    rank[index + 1](RubyAstNode child, int x, int oldIndex |
+      exists(RubyAstNodeParent oldParent |
+        ruby_ast_node_info(child, oldParent, oldIndex, _) and
+        child != parent.(ScopeResolutionMethodCall).getScopeResolution()
+      |
+        oldParent = parent and x = 1
+        or
+        oldParent = parent.(ScopeResolutionMethodCall).getScopeResolution() and x = 0
+      )
+    |
+      child order by x, oldIndex
+    )
```

Comment:
Can you explain what this is doing, and why?

Reply:
A call like `expr::Foo(args) block` used to be represented as `Call{ method: ScopeResolution(expr, Foo), arguments: args, block: block}`. The children of the Call node were:
1. ScopeResolution(expr, Foo)
2. args
3. block
The new parse tree is like `Call{ receiver: expr, method: Foo, arguments: args }` . For this Call the children should be
1. expr
2. Foo
3. args
4. block

To fix the parent child relation for the Call we simply take the children of the `ScopeResolution` first (`x=0`) ordered by `oldIndex` followed by (`x=1`) the other children  of the original Call in the `oldIndex` order. The `rank` is used to ensure the new index is consecutive even if some of the children are missing.